espressif/idf-eclipse-plugin
# IEP-886 Launch Configuration does not save custom build folder path

## Description

Launch configuration was not able to restore the custom build folder after closing and reopening the project. As figured out the Instance Scope preference are cleaned up after the closing of the project, so as a solution I decided to move to store preferences in the launch configuration. It also makes code cleaner and allows us to export the launch configuration with all preferences which are set up there.

Also, as part of PR was some related refactoring of the IDFBuildConfiguration class. The goal of the refactoring was to reduce the complexity of the code. After the refactoring, the build command always picks up the last selected build folder (Previously, it required the restart of the eclipse or editing the cmake file)

Was found and fixed the typo: we were using buildCommand instead of cleanCommand for the cleanCommandTxt.

One last thing, during testing, I found and fixed the root of the issue with the launch configuration, due to which, sometimes the launch configuration could not pick up the selected project and the selected project was always the default after reopening the launch configuration.


Fixes # ([IEP-886](https://jira.espressif.com:8443/browse/IEP-886))

## Type of change

Please delete options that are not relevant.
- Bug fix (non-breaking change which fixes an issue)


## How has this been tested?

Test 1:
- add a custom build folder, close-reopen the project -> make sure that the custom build folder hasn't changed 
Test 2:
- make sure that the refactoring didn't break the current functionality related to the build, indexer, and other related components. We need to do as many tests as possible by comparing behavior on the master and on this branch 

**Test Configuration**:
* ESP-IDF Version:
* OS (Windows,Linux and macOS):

## Dependent components impacted by this PR:

- Build command
- Launch configuration (Build Tab, Main tab)
- custom build folder

## Checklist
